File: scripts/set_asc_price.py
import os, time, json, urllib.request, urllib.error
import jwt

KEY_ID       = os.environ["ASC_KEY_ID"]
ISSUER_ID    = os.environ["ASC_ISSUER_ID"]
PRIVATE_KEY  = os.environ["ASC_PRIVATE_KEY"]
BUNDLE_ID    = "com.reversepicks.app"
APP_ID       = "6781092173"

PLANS = [
    {"product_id": "reversepicks_monthly", "name": "Monthly", "duration": "ONE_MONTH", "price": "36.99"},
]

def make_token():
    now = int(time.time())
    return jwt.encode(
        {"iss": ISSUER_ID, "iat": now, "exp": now + 1200, "aud": "appstoreconnect-v1"},
        PRIVATE_KEY, algorithm="ES256", headers={"kid": KEY_ID},
    )

def asc(method, path, body=None):
    token = make_token()
    data  = json.dumps(body).encode() if body else None
    req   = urllib.request.Request(
        f"https://api.appstoreconnect.apple.com{path}",
        data=data, method=method,
        headers={"Authorization": f"Bearer {token}", "Content-Type": "application/json"},
    )
    try:
        with urllib.request.urlopen(req) as r:
            return r.status, json.loads(r.read())
    except urllib.error.HTTPError as e:
        return e.code, json.loads(e.read())

def ok(status): return status in (200, 201)

# ── 1. Find or create subscription group ──────────────────────────────────────
print("── Subscription group ───────────────────")
status, resp = asc("GET", f"/v1/apps/{APP_ID}/subscriptionGroups?fields[subscriptionGroups]=referenceName")
groups = resp.get("data", [])
print(f"Existing groups: {len(groups)}")

group_id = None
for g in groups:
    print(f"  Found: {g['attributes']['referenceName']} ({g['id']})")
    group_id = g["id"]  # use first existing

if not group_id:
    print("Creating subscription group…")
    status, resp = asc("POST", "/v1/subscriptionGroups", {
        "data": {
            "type": "subscriptionGroups",
            "attributes": {"referenceName": "Reverse Picks Pro"},
            "relationships": {"app": {"data": {"type": "apps", "id": APP_ID}}},
        }
    })
    if ok(status):
        group_id = resp["data"]["id"]
        print(f"✅ Group created: {group_id}")
    else:
        print(f"❌ Group creation failed (HTTP {status}):", json.dumps(resp, indent=2))
        exit(1)

# ── 2. Find or create each subscription ───────────────────────────────────────
print("\n── Subscriptions ────────────────────────")
status, resp = asc("GET", f"/v1/subscriptionGroups/{group_id}/subscriptions?fields[subscriptions]=productId,name,state")
existing_subs = {s["attributes"]["productId"]: s["id"] for s in resp.get("data", [])}
print(f"Existing subs: {list(existing_subs.keys())}")

sub_ids = {}
for plan in PLANS:
    pid = plan["product_id"]
    if pid in existing_subs:
        sub_ids[pid] = existing_subs[pid]
        print(f"  ✅ Already exists: {pid} ({sub_ids[pid]})")
        continue

    print(f"  Creating {pid}…")
    status, resp = asc("POST", "/v1/subscriptions", {
        "data": {
            "type": "subscriptions",
            "attributes": {
                "productId":             pid,
                "name":                  plan["name"],
                "subscriptionPeriod":    plan["duration"],
                "reviewNote":            "Reverse Picks Pro subscription — soccer player props analytics.",
                "familySharable":        False,
            },
            "relationships": {
                "group": {"data": {"type": "subscriptionGroups", "id": group_id}},
            },
        }
    })
    if ok(status):
        sub_ids[pid] = resp["data"]["id"]
        print(f"  ✅ Created: {pid} ({sub_ids[pid]})")
    else:
        print(f"  ❌ Failed (HTTP {status}):", json.dumps(resp, indent=2))

# ── 3. Add English localization (required before review) ──────────────────────
print("\n── Localizations ────────────────────────")
DESCRIPTIONS = {
    "reversepicks_weekly":  "7-day Pro access. Soccer player props analytics.",
    "reversepicks_monthly": "Monthly Pro access. Soccer player props analytics.",
}
for plan in PLANS:
    pid = plan["product_id"]
    if pid not in sub_ids: continue
    sub_id = sub_ids[pid]

    # check existing
    status, resp = asc("GET", f"/v1/subscriptions/{sub_id}/subscriptionLocalizations")
    locs = resp.get("data", [])
    has_en = any(l["attributes"].get("locale","").startswith("en") for l in locs)
    if has_en:
        print(f"  ✅ {pid}: English localization exists")
        continue

    print(f"  Adding English localization for {pid}…")
    status, resp = asc("POST", "/v1/subscriptionLocalizations", {
        "data": {
            "type": "subscriptionLocalizations",
            "attributes": {
                "locale":      "en-US",
                "name":        plan["name"],
                "description": DESCRIPTIONS[pid],
            },
            "relationships": {
                "subscription": {"data": {"type": "subscriptions", "id": sub_id}},
            },
        }
    })
    if ok(status):
        print(f"  ✅ Localization added for {pid}")
    else:
        print(f"  ❌ Localization failed (HTTP {status}):", json.dumps(resp, indent=2))

# ── 4. Set prices ─────────────────────────────────────────────────────────────
print("\n── Prices ───────────────────────────────")
for plan in PLANS:
    pid = plan["product_id"]
    if pid not in sub_ids: continue
    sub_id = sub_ids[pid]
    target_price = plan["price"]

    print(f"  Fetching price points for {pid}…")
    all_pps = []
    next_url = f"/v1/subscriptions/{sub_id}/pricePoints?filter[territory]=USA&fields[subscriptionPricePoints]=customerPrice,proceeds&limit=200"
    while next_url:
        # next_url may be a full URL or a path
        path = next_url if next_url.startswith("/") else "/" + next_url.split("appstoreconnect.apple.com", 1)[-1]
        status, resp = asc("GET", path)
        all_pps.extend(resp.get("data", []))
        raw_next = resp.get("links", {}).get("next")
        next_url = raw_next if raw_next and raw_next != next_url else None
    pps = all_pps
    print(f"  Total price points fetched: {len(pps)}")
    pp = next((p for p in pps if p["attributes"]["customerPrice"] == target_price), None)
    if not pp:
        available = sorted([p["attributes"]["customerPrice"] for p in pps], key=float)
        print(f"  ❌ ${target_price} not found. Sample (highest 10): {available[-10:]}")
        continue
    print(f"  ✅ Found ${target_price} price point: {pp['id']}")

    from datetime import date, timedelta
    start_date = (date.today() + timedelta(days=3)).isoformat()
    status, resp = asc("POST", "/v1/subscriptionPrices", {
        "data": {
            "type": "subscriptionPrices",
            "attributes": {"preserveCurrentPrice": False, "startDate": start_date},
            "relationships": {
                "subscription":           {"data": {"type": "subscriptions",           "id": sub_id}},
                "subscriptionPricePoint": {"data": {"type": "subscriptionPricePoints", "id": pp["id"]}},
            },
        }
    })
    if ok(status):
        print(f"  ✅ {pid} → ${target_price} (effective {start_date})")
    else:
        print(f"  ❌ Price failed (HTTP {status}):", json.dumps(resp, indent=2))

print("\n✅ Done")

File: scripts/setAscPrice.mjs
// App Store Connect API — set reversepicks_weekly to $12.99
// Uses only Node.js built-ins (crypto + https) — no extra packages needed.

import crypto from 'crypto';
import https from 'https';

const KEY_ID     = process.env.ASC_KEY_ID;
const ISSUER_ID  = process.env.ASC_ISSUER_ID;
const PRIVATE_KEY = process.env.ASC_PRIVATE_KEY;

if (!KEY_ID || !ISSUER_ID || !PRIVATE_KEY) {
  console.error('Missing ASC_KEY_ID / ASC_ISSUER_ID / ASC_PRIVATE_KEY');
  process.exit(1);
}

// ── JWT ───────────────────────────────────────────────────────────────────────
function makeJwt() {
  const header  = { alg: 'ES256', kid: KEY_ID, typ: 'JWT' };
  const now     = Math.floor(Date.now() / 1000);
  const payload = { iss: ISSUER_ID, iat: now, exp: now + 1200, aud: 'appstoreconnect-v1' };

  const b64 = (obj) => Buffer.from(JSON.stringify(obj)).toString('base64url');
  const unsigned = `${b64(header)}.${b64(payload)}`;

  const sign = crypto.createSign('SHA256');
  sign.update(unsigned);
  const keyObj = crypto.createPrivateKey({ key: PRIVATE_KEY, format: 'pem' });
  const sigDer = sign.sign({ key: keyObj, dsaEncoding: 'ieee-p1363' });
  const sig = sigDer.toString('base64url');
  return `${unsigned}.${sig}`;
}

// ── HTTPS helper ──────────────────────────────────────────────────────────────
function asc(method, path, body) {
  return new Promise((resolve, reject) => {
    const token = makeJwt();
    const data  = body ? JSON.stringify(body) : null;
    const opts  = {
      hostname: 'api.appstoreconnect.apple.com',
      path,
      method,
      headers: {
        Authorization: `Bearer ${token}`,
        'Content-Type': 'application/json',
        ...(data ? { 'Content-Length': Buffer.byteLength(data) } : {}),
      },
    };
    const req = https.request(opts, (res) => {
      let raw = '';
      res.on('data', (c) => (raw += c));
      res.on('end', () => {
        try { resolve({ status: res.statusCode, body: JSON.parse(raw) }); }
        catch { resolve({ status: res.statusCode, body: raw }); }
      });
    });
    req.on('error', reject);
    if (data) req.write(data);
    req.end();
  });
}

// ── Main ──────────────────────────────────────────────────────────────────────
async function main() {
  // 1. Find the app
  console.log('Looking up apps…');
  const apps = await asc('GET', '/v1/apps?filter[bundleId]=com.reversepicks.app&fields[apps]=name,bundleId', null);
  const appData = apps.body?.data;
  if (!appData?.length) {
    console.error('App not found. Response:', JSON.stringify(apps.body, null, 2));
    process.exit(1);
  }
  const appId = appData[0].id;
  console.log(`✅ App: ${appData[0].attributes?.name} (${appId})`);

  // 2. Find subscription groups
  console.log('\nLooking up subscription groups…');
  const groups = await asc('GET', `/v1/apps/${appId}/subscriptionGroups?fields[subscriptionGroups]=referenceName`, null);
  const groupData = groups.body?.data ?? [];
  console.log(`Found ${groupData.length} group(s)`);
  if (!groupData.length) {
    console.log('No subscription groups found yet — you need to create them in App Store Connect first.');
    console.log('Tip: go to appstoreconnect.apple.com → your app → Monetization → Subscriptions → create a group.');
    process.exit(0);
  }

  // 3. Find reversepicks_weekly subscription in any group
  let weeklySubId = null;
  for (const group of groupData) {
    const subs = await asc('GET', `/v1/subscriptionGroups/${group.id}/subscriptions?fields[subscriptions]=productId,name,state`, null);
    const subData = subs.body?.data ?? [];
    console.log(`  Group "${group.attributes?.referenceName}": ${subData.length} subscription(s)`);
    for (const s of subData) {
      console.log(`    → ${s.attributes?.productId} | ${s.attributes?.name} | ${s.attributes?.state}`);
      if (s.attributes?.productId === 'reversepicks_weekly') {
        weeklySubId = s.id;
      }
    }
  }

  if (!weeklySubId) {
    console.log('\n⚠️  reversepicks_weekly not found in App Store Connect yet.');
    console.log('You need to create it first in the App Store Connect dashboard.');
    console.log('Once created, re-run this script to set the $12.99 price.');
    process.exit(0);
  }

  console.log(`\n✅ Found reversepicks_weekly: ${weeklySubId}`);

  // 4. List existing prices
  console.log('Checking existing prices…');
  const existingPrices = await asc('GET', `/v1/subscriptions/${weeklySubId}/prices?include=subscriptionPricePoint&limit=5`, null);
  const priceData = existingPrices.body?.data ?? [];
  console.log(`Found ${priceData.length} existing price(s)`);

  // 5. Get the price point for $12.99 USD
  console.log('\nFetching price points for USD…');
  const ppRes = await asc('GET', `/v1/subscriptions/${weeklySubId}/pricePoints?filter[territory]=USA&fields[subscriptionPricePoints]=customerPrice,proceeds`, null);
  const pricePoints = ppRes.body?.data ?? [];
  const target = pricePoints.find(pp => pp.attributes?.customerPrice === '12.99');
  if (!target) {
    console.log('$12.99 price point not found. Available USD prices:');
    pricePoints.slice(0, 10).forEach(pp => console.log(`  $${pp.attributes?.customerPrice}`));
    // Try $12.99 by finding closest
    const sorted = pricePoints
      .filter(pp => pp.attributes?.customerPrice)
      .sort((a, b) => Math.abs(parseFloat(a.attributes.customerPrice) - 12.99) - Math.abs(parseFloat(b.attributes.customerPrice) - 12.99));
    console.log(`\nClosest match: $${sorted[0]?.attributes?.customerPrice} (id=${sorted[0]?.id})`);
    process.exit(1);
  }
  console.log(`✅ Found $12.99 price point: ${target.id}`);

  // 6. Set the price
  console.log('\nSetting price to $12.99…');
  const setRes = await asc('POST', '/v1/subscriptionPrices', {
    data: {
      type: 'subscriptionPrices',
      attributes: { preserveCurrentPrice: false, startDate: null },
      relationships: {
        subscription: { data: { type: 'subscriptions', id: weeklySubId } },
        subscriptionPricePoint: { data: { type: 'subscriptionPricePoints', id: target.id } },
      },
    },
  });

  if (setRes.status === 201 || setRes.status === 200) {
    console.log('✅ Price set to $12.99 successfully!');
  } else {
    console.error('❌ Failed to set price:', JSON.stringify(setRes.body, null, 2));
  }
}

main().catch(console.error);
